File: yaml/tests/R/validate.R
# Conformance validation for derivation specifications.
#
# The R counterpart of tests/python/validate.py. Both implementations must
# accept and reject exactly the same specifications. NOTHING CHECKS THAT YET:
# there is no parity harness, and the two are known to disagree.

yaml_dir <- function() {
  normalizePath(file.path(dirname(sys.frame(1)$ofile %||% "."), "..", ".."),
                mustWork = FALSE)
}

`%||%` <- function(a, b) if (is.null(a)) b else a

YAML_DIR <- Sys.getenv("YAMAA_YAML_DIR", unset = normalizePath(
  file.path(dirname(normalizePath(sys.frame(1)$ofile %||% ".", mustWork = FALSE)), "..", ".."),
  mustWork = FALSE
))

load_design <- function(dir = YAML_DIR) {
  list(
    schema = yaml::read_yaml(file.path(dir, "schema.yaml")),
    ops    = yaml::read_yaml(file.path(dir, "operations.yaml"))$operations,
    exc    = yaml::read_yaml(file.path(dir, "exceptions.yaml"))$exceptions
  )
}

# R006 scalar resolution -----------------------------------------------------
# The R yaml package resolves bare Y/N/yes/no/on/off to logical. Every reader
# here must therefore force character, and specifications must not rely on it.

read_yaml_1_2 <- function(path) {
  yaml::read_yaml(path, handlers = list(
    bool_yes = function(x) x,
    bool_no  = function(x) x
  ))
}

scalar_resolution_errors <- function(path) {
  lines <- readLines(path, warn = FALSE)
  pat <- "^\\s*(-\\s*)?[A-Za-z_][A-Za-z0-9_.]*:\\s+(Y|N|yes|no|on|off|Yes|No|On|Off|YES|NO)\\s*$"
  hit <- grep(pat, lines)
  if (!length(hit)) return(character())
  sprintf("%s:%d: unquoted %s resolves as boolean in the R yaml package and/or PyYAML (R006)",
          path, hit, trimws(sub(".*:\\s+", "", lines[hit])))
}

fold_ascii <- function(s) {
  chartype <- strsplit(as.character(s), "")[[1]]
  paste(ifelse(chartype >= "A" & chartype <= "Z", tolower(chartype), chartype),
        collapse = "")
}

# R004 predicate identifiers. Kept deliberately simple: enough to check which
# output columns a predicate names, not a full parser.
predicate_vars <- function(text) {
  kw <- c("AND", "OR", "NOT", "IS", "NULL", "IN", "TRUE", "FALSE")
  stripped <- gsub("'([^']|'')*'", " ", text)
  toks <- regmatches(stripped, gregexpr("[A-Za-z_][A-Za-z0-9_.]*", stripped))[[1]]
  toks[!(toupper(toks) %in% kw)]
}


arg_spec <- function(entry) {
  out <- list()
  for (a in entry$arguments %||% list()) out[[names(a)[1]]] <- a[[1]]
  out
}

as_list_of <- function(x) {
  if (is.null(x)) return(list())
  if (!is.null(names(x))) list(x) else x
}

# R007/R008 registry invariants ----------------------------------------------

check_registries <- function(d = load_design()) {
  errs <- character()
  for (lbl in c("operation", "exception")) {
    reg <- if (lbl == "operation") d$ops else d$exc
    for (nm in names(reg)) {
      entry <- reg[[nm]]
      if (!nzchar(trimws(entry$semantics %||% ""))) {
        errs <- c(errs, sprintf("registry: %s '%s' has no semantics", lbl, nm))
      }
      for (arg in names(arg_spec(entry))) {
        desc <- arg_spec(entry)[[arg]]
        if (is.null(desc$default)) next
        if (isTRUE(desc$required)) {
          errs <- c(errs, sprintf("registry: %s.%s has a default on a required argument",
                                  nm, arg))
        }
        if (!is.null(desc$values) && !(desc$default %in% desc$values)) {
          errs <- c(errs, sprintf("registry: %s.%s default '%s' is not in [%s]",
                                  nm, arg, desc$default, paste(desc$values, collapse = ", ")))
        }
        ok <- switch(desc$type %||% "",
                     bool = is.logical(desc$default),
                     int = is.numeric(desc$default),
                     str = is.character(desc$default),
                     float = is.numeric(desc$default),
                     TRUE)
        if (!ok) {
          errs <- c(errs, sprintf("registry: %s.%s default does not satisfy type '%s'",
                                  nm, arg, desc$type))
        }
      }
    }
  }
  errs
}

# Specification validation ----------------------------------------------------

validate_spec <- function(path, d = load_design()) {
  errs <- scalar_resolution_errors(path)
  spec <- read_yaml_1_2(path)
  datasets <- names(spec$datasets %||% list())

  fail <- function(...) errs <<- c(errs, sprintf(...))

  closed <- function(val, cls, where) {
    allowed <- vapply(d$schema[[cls]], function(e) names(e)[1], "")
    for (k in names(val)) {
      if (!(k %in% allowed)) fail("%s: undeclared field '%s' for %s", where, k, cls)
    }
    for (e in d$schema[[cls]]) {
      nm <- names(e)[1]
      if (isTRUE(e[[1]]$required) && is.null(val[[nm]])) {
        fail("%s: missing required field '%s'", where, nm)
      }
    }
  }

  action <- function(act, where, reg, lbl) {
    if (length(act) != 1L) {
      fail("%s: action_class requires size 1", where); return(NULL)
    }
    nm <- names(act)[1]
    if (!(nm %in% names(reg))) {
      fail("%s: '%s' is not in the %s registry", where, nm, lbl); return(NULL)
    }
    entry <- reg[[nm]]
    args <- act[[1]] %||% list()
    spc <- arg_spec(entry)
    for (a in names(args)) {
      if (!(a %in% names(spc))) fail("%s.%s: unknown argument '%s'", where, nm, a)
    }
    for (a in names(spc)) {
      desc <- spc[[a]]
      if (isTRUE(desc$required) && is.null(args[[a]])) {
        fail("%s.%s: missing required argument '%s'", where, nm, a)
      }
      if (is.null(args[[a]])) next
      if (!is.null(desc$values) && !(args[[a]] %in% desc$values)) {
        fail("%s.%s.%s: '%s' is not one of [%s]", where, nm, a, args[[a]],
             paste(desc$values, collapse = ", "))
      }
      if (identical(desc$type, "dataset_id") && !(args[[a]] %in% datasets)) {
        fail("%s.%s.%s: '%s' is not a declared dataset", where, nm, a, args[[a]])
      }
    }
    if (identical(nm, "mapping") && identical(args$case_sensitive, FALSE)) {
      folded <- vapply(names(args$dict), fold_ascii, "")
      dup <- folded[duplicated(folded)]
      for (f in unique(dup)) {
        fail("%s.mapping.dict: keys fold to the same value under case_sensitive: false",
             where)
      }
    }
    entry
  }

  check_predicate <- function(text, where, label, declared) {
    for (n in predicate_vars(text)) {
      if (grepl(".", n, fixed = TRUE)) {
        fail("%s.%s: '%s' is qualified; a predicate over output rows names output columns (R001)",
             where, label, n)
      } else if (!(n %in% declared)) {
        fail("%s.%s: '%s' is not a declared output column (R001)", where, label, n)
      }
    }
  }

  derivation <- function(dv, where, driver) {
    closed(dv, "derivation_class", where)
    if (!is.null(dv$where)) check_predicate(dv$where, where, "where", declared)
    src <- dv$source
    if (!is.null(src) && grepl(".", src, fixed = TRUE)) {
      q <- sub("\\..*$", "", src)
      if (!(q %in% datasets)) {
        fail("%s.source: '%s' is not a declared dataset (R002)", where, q)
      }
    }
    joins <- !is.null(src) && grepl(".", src, fixed = TRUE) &&
      sub("\\..*$", "", src) != (driver %||% "")
    if (!is.null(dv$filter) && !joins) {
      fail("%s.filter: derivation performs no join (R003)", where)
    }

    ops <- as_list_of(dv$operations)
    entries <- list()
    for (i in seq_along(ops)) {
      e <- action(ops[[i]], sprintf("%s.operations[%d]", where, i - 1L), d$ops, "operation")
      entries[[i]] <- e
      if (!is.null(e) && identical(e$kind, "aggregate") &&
          !((i == 1L && joins) || !is.null(dv$group_by))) {
        fail("%s.operations[%d]: aggregate outside its two legal positions (R007)",
             where, i - 1L)
      }
    }
    seeded <- !is.null(dv$source) || !is.null(dv$literal)
    if (length(ops) && !is.null(entries[[1]]) &&
        identical(entries[[1]]$seed, "required") && !seeded) {
      fail("%s: leading operation declares seed: required but there is no source or literal (R007)",
           where)
    }
    if (!length(ops) && !seeded) fail("%s: no source, literal, or operations (R004)", where)

    stages <- character()
    exs <- as_list_of(dv$exception)
    for (i in seq_along(exs)) {
      e <- action(exs[[i]], sprintf("%s.exception[%d]", where, i - 1L), d$exc, "exception")
      if (is.null(e)) next
      nm <- names(exs[[i]])[1]
      if (e$stage %in% stages) {
        fail("%s.exception[%d].%s: stage '%s' already bound (R008)", where, i - 1L, nm, e$stage)
      }
      stages <- c(stages, e$stage)
      if (identical(e$stage, "operation")) {
        n <- sum(vapply(entries, function(x) !is.null(x) && nm %in% (x$raises %||% character()),
                        logical(1)))
        if (n != 1L) {
          fail("%s.exception[%d].%s: bound to %d operations that raise it, requires 1 (R008)",
               where, i - 1L, nm, n)
        }
      }
    }
  }

  declared <- vapply(spec$columns, function(c) c$name, "")
  closed(spec, "root_class", "root")
  if (!is.null(spec$domain) && spec$domain %in% datasets) {
    fail("root.datasets: '%s' is also the output domain (R002)", spec$domain)
  }
  rows <- spec$rows %||% list()
  if (!length(rows) && is.null(spec$base)) fail("root: no rows entry and no base (R001)")
  if (!is.null(spec$base) && !(spec$base %in% datasets)) {
    fail("root.base: '%s' is not a declared dataset (R002)", spec$base)
  }

  for (col in spec$columns) {
    closed(col, "column_class", sprintf("columns[%s]", col$name))
    if (!is.null(col$derivation)) {
      derivation(col$derivation, sprintf("columns[%s].derivation", col$name), spec$base)
    }
  }
  for (r in rows) {
    closed(r, "row_class", sprintf("rows[%s]", r$id))
    if (!is.null(r$dataset) && !(r$dataset %in% datasets)) {
      fail("rows[%s].dataset: '%s' is not declared (R002)", r$id, r$dataset)
    }
    for (nm in names(r$derivations)) {
      if (!(nm %in% declared)) fail("rows[%s]: target '%s' is not a declared column", r$id, nm)
      derivation(r$derivations[[nm]], sprintf("rows[%s].%s", r$id, nm),
                 r$dataset %||% spec$base)
    }
  }

  col_derived <- vapply(spec$columns, function(c) !is.null(c$derivation), logical(1))
  col_derived <- declared[col_derived]
  row_targets <- lapply(rows, function(r) names(r$derivations))
  for (nm in declared) {
    covered <- nm %in% col_derived ||
      (length(row_targets) && all(vapply(row_targets, function(t) nm %in% t, logical(1))))
    if (!covered) fail("R005 coverage: '%s' has no column derivation and is missing from a rows entry", nm)
  }
  for (k in spec$keys %||% character()) {
    if (!(k %in% declared)) fail("R005: key '%s' is not a declared column", k)
  }

  errs
}

example_specs <- function(dir = YAML_DIR) {
  sort(Sys.glob(file.path(dir, "examples", "*", "spec.yaml")))
}

File: yaml/tests/python/validate.py
"""Conformance validation for derivation specifications.

Checks a spec against schema.yaml (R006), operations.yaml (R007),
exceptions.yaml (R008), and the structural clauses of R001, R002, R003 and R005.

This validates shape and vocabulary only. Executing a specification and
comparing its output to the expected CSV is engine.py.
"""
from __future__ import annotations

import re
from pathlib import Path

import yaml

from functions import HOST_FUNCTIONS
from predicates import DerivationError, check_syntax, predicate_vars

YAML_DIR = Path(__file__).resolve().parents[2]


def _load(name):
    with open(YAML_DIR / name) as fh:
        return yaml.safe_load(fh)


SCHEMA = _load("schema.yaml")
OPS = _load("operations.yaml")["operations"]
EXC = _load("exceptions.yaml")["exceptions"]
VARPAT = re.compile(SCHEMA["variable"]["pattern"])

# R006 scalar resolution: values PyYAML or the R yaml package resolve to boolean.
# Flagged in raw text because by parse time the information is already lost.
_BOOLISH = re.compile(
    r"^\s*(?:-\s*)?[A-Za-z_][\w.]*:\s+(Y|N|yes|no|on|off|Yes|No|On|Off|YES|NO)\s*$"
)

_SCALARS = {"bool": bool, "int": int, "str": str, "float": (int, float)}


def _fields(cls):
    return [next(iter(e)) for e in SCHEMA[cls]]


def _argspec(entry):
    return {next(iter(a)): next(iter(a.values())) for a in (entry.get("arguments") or [])}


def _as_list(v):
    if v is None:
        return []
    return v if isinstance(v, list) else [v]


def fold_ascii(s):
    """R007 case folding: A-Z to a-z, every other code point untouched."""
    return "".join(chr(ord(c) + 32) if "A" <= c <= "Z" else c for c in str(s))


def check_registries():
    """R007/R008 invariants on the registry documents themselves."""
    errs = []
    for label, reg in (("operation", OPS), ("exception", EXC)):
        for name, entry in reg.items():
            if not (entry.get("semantics") or "").strip():
                errs.append(f"registry: {label} {name!r} has no semantics")
            for arg, desc in _argspec(entry).items():
                if "default" not in desc:
                    continue
                if desc.get("required"):
                    errs.append(f"registry: {name}.{arg} has a default on a required argument")
                if "values" in desc and desc["default"] not in desc["values"]:
                    errs.append(
                        f"registry: {name}.{arg} default {desc['default']!r} "
                        f"is not in {desc['values']}"
                    )
                want = _SCALARS.get(desc.get("type"))
                if want and not isinstance(desc["default"], want):
                    errs.append(
                        f"registry: {name}.{arg} default {desc['default']!r} "
                        f"does not satisfy type {desc['type']!r}"
                    )
    return errs


class _Validator:
    def __init__(self, spec, raw):
        self.spec = spec
        self.raw = raw
        self.datasets = spec.get("datasets") or {}
        self.declared = {c["name"] for c in spec.get("columns") or []}
        self.errs = []

    def fail(self, msg):
        self.errs.append(msg)

    # ---- R006 -----------------------------------------------------------

    def closed(self, val, cls, where):
        allowed = set(_fields(cls))
        for key in val:
            if key not in allowed:
                self.fail(f"{where}: undeclared field {key!r} for {cls}")
        for entry in SCHEMA[cls]:
            name, desc = next(iter(entry.items()))
            if desc.get("required") and name not in val:
                self.fail(f"{where}: missing required field {name!r}")

    def scalars(self, path):
        for n, line in enumerate(self.raw.splitlines(), 1):
            m = _BOOLISH.match(line)
            if m:
                self.fail(
                    f"{path}:{n}: unquoted {m.group(1)!r} resolves as boolean in the R "
                    f"yaml package and/or PyYAML (R006 scalar resolution)"
                )

    # ---- R007 / R008 ----------------------------------------------------

    def action(self, act, where, registry, label):
        """action_class: a size-1 mapping of registered name to named arguments."""
        if not isinstance(act, dict) or len(act) != 1:
            self.fail(f"{where}: action_class requires size 1, got {act!r}")
            return None, None
        name, args = next(iter(act.items()))
        if name not in registry:
            self.fail(f"{where}: {name!r} is not in the {label} registry")
            return None, None
        entry = registry[name]
        args = args or {}
        spec = _argspec(entry)
        for arg in args:
            if arg not in spec:
                self.fail(
                    f"{where}.{name}: unknown argument {arg!r} "
                    f"(registered: {sorted(spec) or 'none'})"
                )
        for arg, desc in spec.items():
            if desc.get("required") and arg not in args:
                self.fail(f"{where}.{name}: missing required argument {arg!r}")
            if arg not in args:
                continue
            if "values" in desc and args[arg] not in desc["values"]:
                self.fail(
                    f"{where}.{name}.{arg}: {args[arg]!r} is not one of {desc['values']}"
                )
            if desc.get("type") == "dataset_id" and args[arg] not in self.datasets:
                self.fail(
                    f"{where}.{name}.{arg}: {args[arg]!r} is not a declared dataset "
                    f"(declared: {sorted(self.datasets)})"
                )
        for arg, desc in spec.items():
            if desc.get("type") != "sql" or arg not in args:
                continue
            self.predicate(args[arg], f"{where}.{name}", arg)
        for arg in args:
            for item in (args[arg] if isinstance(args[arg], list) else []):
                if isinstance(item, dict) and isinstance(item.get("when"), str):
                    self.predicate(item["when"], f"{where}.{name}.{arg}", "when")
        for arg, val in args.items():
            self.no_nested_action(val, f"{where}.{name}.{arg}")
        if name == "call" and isinstance(args.get("function"), str):
            if args["function"] not in HOST_FUNCTIONS:
                self.fail(
                    f"{where}.call.function: {args['function']!r} is not in the host "
                    f"function library (available: {sorted(HOST_FUNCTIONS)}) (R007)"
                )
        if name == "mapping" and args.get("case_sensitive") is False:
            seen = {}
            for key in args.get("dict") or {}:
                folded = fold_ascii(key)
                if folded in seen:
                    self.fail(
                        f"{where}.mapping.dict: keys {seen[folded]!r} and {key!r} fold "
                        f"to the same value under case_sensitive: false"
                    )
                seen[folded] = key
        return name, entry

    def no_nested_action(self, val, where):
        """An argument holds values, never an operation. R007: `action_argument`
        has no `action_class` member, so a nested operation is a silent no-op
        that lands the mapping itself in the output."""
        if isinstance(val, dict):
            if len(val) == 1:
                only = next(iter(val))
                if only != "source" and (only in OPS or only in EXC):
                    self.fail(
                        f"{where}: {only!r} is an operation nested where a value is "
                        f"expected; arguments cannot hold operations (R007)"
                    )
            for k, sub in val.items():
                self.no_nested_action(sub, f"{where}.{k}")
        elif isinstance(val, list):
            for i, sub in enumerate(val):
                self.no_nested_action(sub, f"{where}[{i}]")

    # ---- derivations ----------------------------------------------------

    def predicate(self, text, where, label):
        """R001: a predicate may only name declared output columns."""
        try:
            check_syntax(text)
            names = predicate_vars(text)
        except DerivationError as exc:
            self.fail(f"{where}.{label}: {exc}")
            return
        for n in names:
            if "." in n:
                self.fail(f"{where}.{label}: {n!r} is qualified; a predicate over "
                          f"output rows names output columns (R001)")
            elif n not in self.declared:
                self.fail(f"{where}.{label}: {n!r} is not a declared output column (R001)")

    def derivation(self, d, where, driver):
        self.closed(d, "derivation_class", where)
        if d.get("where"):
            self.predicate(d["where"], where, "where")

        src = d.get("source")
        if src is not None and not VARPAT.match(str(src)):
            self.fail(f"{where}.source: {src!r} fails the variable pattern")
        qualifier = str(src).split(".")[0] if src and "." in str(src) else None
        if qualifier is not None and qualifier not in self.datasets:
            self.fail(
                f"{where}.source: {qualifier!r} is not a declared dataset "
                f"(declared: {sorted(self.datasets)}) (R002)"
            )

        joins = qualifier is not None and qualifier != driver and qualifier in self.datasets
        if d.get("filter") is not None and not isinstance(d["filter"], str):
            self.fail(f"{where}.filter: must be one sql predicate (R004)")
        if "filter" in d and not joins:
            self.fail(f"{where}.filter: derivation performs no join (R003)")

        ops = _as_list(d.get("operations"))
        resolved = []
        for i, op in enumerate(ops):
            name, entry = self.action(
                op, f"{where}.operations[{i}]", OPS, "operation"
            )
            resolved.append((name, entry))
            if entry and entry["kind"] == "aggregate" and not (
                (i == 0 and joins) or "group_by" in d
            ):
                self.fail(
                    f"{where}.operations[{i}].{name}: aggregate outside its two legal "
                    f"positions -- not a leading right-side reduction, no group_by (R007)"
                )

        seeded = "source" in d or "literal" in d
        if ops and resolved[0][1] and resolved[0][1]["seed"] == "required" and not seeded:
            self.fail(
                f"{where}: leading {resolved[0][0]!r} declares seed: required but the "
                f"derivation has no source or literal (R007)"
            )
        if not ops and not seeded:
            self.fail(f"{where}: no source, literal, or operations (R004)")
        if "source" in d and "literal" in d:
            self.fail(f"{where}: both source and literal (R004)")

        stages = {}
        for i, ex in enumerate(_as_list(d.get("exception"))):
            name, entry = self.action(
                ex, f"{where}.exception[{i}]", EXC, "exception"
            )
            if not entry:
                continue
            stage = entry["stage"]
            if stage in stages:
                self.fail(
                    f"{where}.exception[{i}].{name}: stage {stage!r} already bound by "
                    f"{stages[stage]!r} (R008 permits one per stage)"
                )
            stages[stage] = name
            if stage == "operation":
                n = sum(1 for _, e in resolved if e and name in (e.get("raises") or []))
                if n != 1:
                    self.fail(
                        f"{where}.exception[{i}].{name}: bound to {n} operations that "
                        f"raise it, requires exactly 1 (R008)"
                    )

    # ---- entry point ----------------------------------------------------

    def run(self, path):
        self.scalars(path)
        self.closed(self.spec, "root_class", "root")

        domain = self.spec.get("domain")
        if domain in self.datasets:
            self.fail(
                f"root.datasets: {domain!r} is also the output domain; a source dataset "
                f"must not reuse the domain name (R002)"
            )

        rows = self.spec.get("rows") or []
        base = self.spec.get("base")
        if not rows and base is None:
            self.fail("root: no rows entry and no base (R001)")
        if base is not None and base not in self.datasets:
            self.fail(f"root.base: {base!r} is not a declared dataset (R002)")

        declared = {c["name"] for c in self.spec.get("columns", [])}
        for col in self.spec.get("columns", []):
            self.closed(col, "column_class", f"columns[{col.get('name')}]")
            if "derivation" in col:
                self.derivation(
                    col["derivation"], f"columns[{col['name']}].derivation", base
                )

        for row in rows:
            rid = row.get("id")
            self.closed(row, "row_class", f"rows[{rid}]")
            if row.get("filter") is not None and not isinstance(row["filter"], str):
                self.fail(f"rows[{rid}].filter: must be one sql predicate (R004)")
            if "dataset" in row and row["dataset"] not in self.datasets:
                self.fail(f"rows[{rid}].dataset: {row['dataset']!r} is not declared (R002)")
            for name, d in (row.get("derivations") or {}).items():
                if name not in declared:
                    self.fail(f"rows[{rid}]: target {name!r} is not a declared column")
                self.derivation(d, f"rows[{rid}].{name}", row.get("dataset", base))

        col_derived = {
            c["name"] for c in self.spec.get("columns", []) if "derivation" in c
        }
        row_targets = [set(r.get("derivations") or {}) for r in rows]
        for name in sorted(declared):
            covered = name in col_derived or (
                row_targets and all(name in t for t in row_targets)
            )
            if not covered:
                self.fail(
                    f"R005 coverage: {name!r} has no column derivation and is missing "
                    f"from at least one rows entry"
                )
        for key in self.spec.get("keys", []):
            if key not in declared:
                self.fail(f"R005: key {key!r} is not a declared column")

        return self.errs


def validate_spec(path):
    """Return a list of conformance errors for one specification file."""
    path = Path(path)
    raw = path.read_text()
    return _Validator(yaml.safe_load(raw), raw).run(str(path))


def example_specs():
    return sorted((YAML_DIR / "examples").glob("*/spec.yaml"))


if __name__ == "__main__":
    import sys

    failed = check_registries()
    for err in failed:
        print(f"FAIL {err}")
    for spec in example_specs():
        errs = validate_spec(spec)
        print(f"{'FAIL' if errs else 'ok  '} {spec.parent.name}")
        for err in errs:
            print(f"       {err}")
        failed += errs
    sys.exit(1 if failed else 0)
